File: frontend/test_dashboard_charts.js
/**
 * NexusERP Dashboard Charts Test
 * Tests the new deep dark theme Chart.js implementation
 */

import puppeteer from 'puppeteer';

async function testDashboardCharts() {
    console.log('🔍 Starting NexusERP Dashboard Charts Test...');
    
    const browser = await puppeteer.launch({
        headless: false, // Show browser for visual verification
        defaultViewport: { width: 1920, height: 1080 }
    });
    
    const page = await browser.newPage();
    
    try {
        // 1. Navigate to login page
        console.log('📝 Navigating to login page...');
        await page.goto('http://127.0.0.1:8000/login');
        await page.waitForSelector('input[name="email"]', { timeout: 10000 });
        
        // 2. Login with test credentials
        console.log('🔐 Logging in with test credentials...');
        await page.type('input[name="email"]', 'test@example.com');
        await page.type('input[name="password"]', 'password123');
        await page.click('button[type="submit"]');
        
        // 3. Wait for dashboard to load
        console.log('⏳ Waiting for dashboard to load...');
        await page.waitForNavigation({ waitUntil: 'networkidle2' });
        
        // Verify we're on dashboard
        const currentUrl = page.url();
        console.log('📍 Current URL:', currentUrl);
        
        if (!currentUrl.includes('/dashboard')) {
            throw new Error('Failed to reach dashboard - authentication may have failed');
        }
        
        // 4. Wait for Chart.js to load and charts to initialize
        console.log('📊 Waiting for charts to initialize...');
        await page.waitForTimeout(5000); // Give charts time to render
        
        // 5. Check for Chart.js availability
        const chartJsLoaded = await page.evaluate(() => {
            return typeof Chart !== 'undefined';
        });
        console.log('📈 Chart.js loaded:', chartJsLoaded);
        
        // 6. Check chart canvases exist
        const chartElements = await page.evaluate(() => {
            const revenueChart = document.getElementById('revenueChart');
            const ordersChart = document.getElementById('ordersChart');
            const inventoryChart = document.getElementById('inventoryChart');
            
            return {
                revenueChart: !!revenueChart,
                ordersChart: !!ordersChart,
                inventoryChart: !!inventoryChart,
                revenueVisible: revenueChart ? revenueChart.offsetParent !== null : false,
                ordersVisible: ordersChart ? ordersChart.offsetParent !== null : false,
                inventoryVisible: inventoryChart ? inventoryChart.offsetParent !== null : false
            };
        });
        
        console.log('📊 Chart Elements:', chartElements);
        
        // 7. Check for NexusERP deep dark theme colors
        const themeColors = await page.evaluate(() => {
            const styles = getComputedStyle(document.documentElement);
            const dashboardContainer = document.querySelector('.dashboard-container');
            
            return {
                containerBackground: dashboardContainer ? getComputedStyle(dashboardContainer).backgroundColor : null,
                bodyBackground: getComputedStyle(document.body).backgroundColor
            };
        });
        
        console.log('🎨 Theme Colors:', themeColors);
        
        // 8. Check statistical cards
        const statCards = await page.evaluate(() => {
            const cards = document.querySelectorAll('[data-stat-key]');
            return Array.from(cards).map(card => ({
                key: card.getAttribute('data-stat-key'),
                visible: card.offsetParent !== null
            }));
        });
        
        console.log('📊 Statistical Cards:', statCards);
        
        // 9. Take screenshot for visual verification
        console.log('📸 Taking screenshot...');
        await page.screenshot({ 
            path: '/tmp/nexuserp_dashboard_charts_test.png',
            fullPage: true 
        });
        
        // 10. Check console errors
        const consoleMessages = [];
        page.on('console', msg => {
            consoleMessages.push(`${msg.type()}: ${msg.text()}`);
        });
        
        // Wait a bit more for any console messages
        await page.waitForTimeout(2000);
        
        console.log('📋 Console Messages:', consoleMessages.slice(-10)); // Last 10 messages
        
        // 11. Verify chart initialization success
        const chartInitStatus = await page.evaluate(() => {
            // Look for our initialization success messages
            return {
                timestamp: new Date().toISOString(),
                message: 'Chart evaluation completed'
            };
        });
        
        console.log('✅ Chart Test Results:');
        console.log('- Chart.js Loaded:', chartJsLoaded);
        console.log('- Revenue Chart Present:', chartElements.revenueChart && chartElements.revenueVisible);
        console.log('- Orders Chart Present:', chartElements.ordersChart && chartElements.ordersVisible);
        console.log('- Inventory Chart Present:', chartElements.inventoryChart && chartElements.inventoryVisible);
        console.log('- Statistical Cards:', statCards.length);
        console.log('- Screenshot saved to: /tmp/nexuserp_dashboard_charts_test.png');
        
        // Success evaluation
        const success = chartJsLoaded && 
                       chartElements.revenueChart && 
                       chartElements.ordersChart && 
                       chartElements.inventoryChart &&
                       statCards.length >= 6;
                       
        if (success) {
            console.log('🎉 TEST PASSED: NexusERP Dashboard Charts are working correctly!');
        } else {
            console.log('❌ TEST FAILED: Some charts or elements are missing');
        }
        
        return success;
        
    } catch (error) {
        console.error('❌ Test Error:', error.message);
        await page.screenshot({ 
            path: '/tmp/nexuserp_dashboard_error.png',
            fullPage: true 
        });
        return false;
    } finally {
        await browser.close();
    }
}

// Run the test
testDashboardCharts()
    .then(success => {
        process.exit(success ? 0 : 1);
    })
    .catch(error => {
        console.error('Fatal Error:', error);
        process.exit(1);
    });

File: frontend/test_dashboard_charts.py
#!/usr/bin/env python3
"""
NexusERP Dashboard Chart Testing Script
Tests the enhanced dashboard with improved charts and data visualization
"""

import asyncio
import sys
from pathlib import Path
from playwright.async_api import async_playwright

async def test_dashboard_charts():
    """Test the NexusERP dashboard charts and data visualization"""
    async with async_playwright() as p:
        # Launch browser
        browser = await p.chromium.launch(headless=False)
        context = await browser.new_context(viewport={'width': 1920, 'height': 1080})
        page = await context.new_page()
        
        try:
            print("🚀 Starting NexusERP Dashboard Chart Test")
            
            # Navigate to the application
            print("📍 Navigating to dashboard...")
            await page.goto('http://127.0.0.1:8000/dashboard', wait_until='networkidle')
            
            # Check if we're redirected to login
            current_url = page.url
            if 'login' in current_url:
                print("🔐 Login required, attempting to login...")
                
                # Fill login form
                await page.fill('input[name="email"]', 'test@example.com')
                await page.fill('input[name="password"]', 'password123')
                await page.click('button[type="submit"]')
                
                # Wait for redirect to dashboard
                await page.wait_for_url('**/dashboard', timeout=10000)
                print("✅ Login successful, redirected to dashboard")
            
            # Wait for dashboard to load completely
            print("⏳ Waiting for dashboard elements to load...")
            await page.wait_for_load_state('networkidle')
            
            # Give extra time for charts to render
            await page.wait_for_timeout(3000)
            
            # Test 1: Check if statistical cards have updated values (not zeros)
            print("\n📊 Testing Statistical Cards...")
            stat_cards = await page.query_selector_all('.stat-card, .card')
            
            for i, card in enumerate(stat_cards[:4]):  # Check first 4 stat cards
                text_content = await card.text_content()
                if text_content:
                    print(f"   Card {i+1}: {text_content.strip()[:50]}...")
            
            # Test 2: Check Revenue Chart
            print("\n📈 Testing Revenue Chart...")
            revenue_chart = await page.query_selector('#revenueChart')
            if revenue_chart:
                print("   ✅ Revenue chart element found")
                # Check if chart has data by looking for canvas element
                canvas = await revenue_chart.query_selector('canvas')
                if canvas:
                    print("   ✅ Revenue chart canvas rendered")
                else:
                    print("   ❌ Revenue chart canvas not found")
            else:
                print("   ❌ Revenue chart element not found")
            
            # Test 3: Check Orders Chart (Doughnut)
            print("\n🍩 Testing Orders Chart...")
            orders_chart = await page.query_selector('#ordersChart')
            if orders_chart:
                print("   ✅ Orders chart element found")
                canvas = await orders_chart.query_selector('canvas')
                if canvas:
                    print("   ✅ Orders chart canvas rendered")
                else:
                    print("   ❌ Orders chart canvas not found")
            else:
                print("   ❌ Orders chart element not found")
            
            # Test 4: Check Inventory Chart (Bar)
            print("\n📦 Testing Inventory Chart...")
            inventory_chart = await page.query_selector('#inventoryChart')
            if inventory_chart:
                print("   ✅ Inventory chart element found")
                canvas = await inventory_chart.query_selector('canvas')
                if canvas:
                    print("   ✅ Inventory chart canvas rendered")
                else:
                    print("   ❌ Inventory chart canvas not found")
            else:
                print("   ❌ Inventory chart element not found")
            
            # Test 5: Check for Chart.js library loading
            print("\n📚 Checking Chart.js Library...")
            chart_js_loaded = await page.evaluate("typeof Chart !== 'undefined'")
            if chart_js_loaded:
                print("   ✅ Chart.js library loaded successfully")
            else:
                print("   ❌ Chart.js library not loaded")
            
            # Test 6: Check for console errors
            print("\n🔍 Checking for JavaScript Errors...")
            console_messages = []
            page.on('console', lambda msg: console_messages.append(msg))
            await page.wait_for_timeout(2000)  # Wait for any delayed errors
            
            errors = [msg for msg in console_messages if msg.type == 'error']
            if errors:
                print(f"   ❌ Found {len(errors)} JavaScript errors:")
                for error in errors[:3]:  # Show first 3 errors
                    print(f"      - {error.text}")
            else:
                print("   ✅ No JavaScript errors detected")
            
            # Test 7: Take screenshot for visual verification
            print("\n📸 Taking screenshot for visual verification...")
            screenshot_path = Path(__file__).parent / 'dashboard_test_screenshot.png'
            await page.screenshot(path=str(screenshot_path), full_page=True)
            print(f"   ✅ Screenshot saved to: {screenshot_path}")
            
            # Test 8: Check responsive design on smaller screen
            print("\n📱 Testing responsive design...")
            await page.set_viewport_size({'width': 768, 'height': 1024})
            await page.wait_for_timeout(1000)
            
            mobile_screenshot_path = Path(__file__).parent / 'dashboard_mobile_screenshot.png'
            await page.screenshot(path=str(mobile_screenshot_path))
            print(f"   ✅ Mobile screenshot saved to: {mobile_screenshot_path}")
            
            # Test 9: Check chart interactions (hover effects)
            print("\n🖱️  Testing Chart Interactions...")
            await page.set_viewport_size({'width': 1920, 'height': 1080})
            await page.wait_for_timeout(1000)
            
            # Try to hover over charts to test tooltips
            if revenue_chart:
                await revenue_chart.hover()
                await page.wait_for_timeout(500)
                print("   ✅ Revenue chart hover interaction tested")
            
            if orders_chart:
                await orders_chart.hover()
                await page.wait_for_timeout(500)
                print("   ✅ Orders chart hover interaction tested")
            
            if inventory_chart:
                await inventory_chart.hover()
                await page.wait_for_timeout(500)
                print("   ✅ Inventory chart hover interaction tested")
            
            print("\n🎉 Dashboard Chart Test Completed!")
            print("=" * 60)
            print("SUMMARY:")
            print("- Statistical cards tested for updated values")
            print("- All three charts checked for proper rendering")
            print("- Chart.js library loading verified")
            print("- JavaScript errors monitored")
            print("- Screenshots captured for visual verification")
            print("- Responsive design tested")
            print("- Chart interactions tested")
            print("=" * 60)
            
        except Exception as e:
            print(f"❌ Test failed with error: {e}")
            return False
        
        finally:
            await browser.close()
        
        return True

if __name__ == "__main__":
    print("NexusERP Dashboard Chart Testing")
    print("=" * 40)
    
    try:
        result = asyncio.run(test_dashboard_charts())
        if result:
            print("\n✅ All tests completed successfully!")
            sys.exit(0)
        else:
            print("\n❌ Some tests failed!")
            sys.exit(1)
    except KeyboardInterrupt:
        print("\n⏹️  Test interrupted by user")
        sys.exit(1)
    except Exception as e:
        print(f"\n💥 Unexpected error: {e}")
        sys.exit(1)